### ISSUE TITLE
[umbrella issue]: decoupling from k8s where k8s isn't the best option

### ISSUE THREAD

**Issue by krancour** [brigadecore/brigade#1034]
While it's enormously useful that Brigade can schedule work in a Kubernetes cluster in response to (often external) events, many of Brigade's _other_ uses of Kubernetes are incidental / implementation details. In many such cases, superior options to Kubernetes may exist, new benefits might be derived from decoupling from Kubernetes-- or both.

This is an umbrella issue to track a number of discrete issues (mostly proposals) that relate to this theme and involve major architectural changes (i.e. breaking changes) that we may wish to implement in the (TDB) 2.0 time frame.

* https://github.com/brigadecore/brigade/issues/1035
* https://github.com/brigadecore/brigade/issues/1036
* https://github.com/brigadecore/brigade/issues/1037

__Edit:__ After a lot more thought, I have realized that we _do_ derive certain benefits from using k8s as a datastore (particularly in the case of project configuration). Namely, it allows us to lean on k8s for authentication and authorization instead of implementing user management and RBAC ourselves. I wanted to take a moment to acknowledge this... then argue that I am still in favor of some big changes here.

__As I see it, the fundamental question we have to answer going into 2.0 is this:__

Do we want Kubernetes to (continue to) completely permeate Brigade's design? Do we want access to a cluster and a decent level of k8s knowledge to be a prerequisite for using Brigade? Or, alternatively, do we want to not pre-suppose those things and, in the most generic sense, have Brigade be a platform for receiving events and executing workflows in response, with the fact that the workflows execute _in_ a Kubernetes cluster being a mere implementation detail?

**Comment by technosophos** [comment]
I could see trying to isolate all of the Kubernetes-specific code to a "driver layer" and make sure that we could easily swap out.

One of the projects I had initially wanted to try out with Brigade WAAAAY back was to see if I could run it on Nomad. Of course, practical design decisions at that time redirected us to be 💯 Kubernetes and I never really gave it a shot... but even today I'd have to say that Nomad (and other orchestrators) should be able _in theory_ to run the same sort of things at the same level of abstraction as Kubernetes.

One mistake I think I have made over the life of Brigade is to let some of the Kubernetes objects "leak" into the JavaScript layer. But as I look, even those are not exactly Kubernetes specific.

Bottom line: I'd love to give this a shot.

**Comment by danmx** [comment]
It make sense to not rely on k8s when it’s not the best tool for the job but to make brigade platform agnostic, I think it’s not the right time. A lot of CI and Workflow projects are moving from other platforms to k8s and have to spend time adjusting or changing their architecture to support it.

For example:
Airflow supports k8s for its workers but Google is still building proper [Airflow operator](https://github.com/GoogleCloudPlatform/airflow-operator) to fully run on k8s.
[Kubeflow](https://www.kubeflow.org/docs/pipelines/overview/pipelines-overview/) leverages K8s native workflow engine Argo and both solutions are wildly successful but they had to build an interpreter that will translate Python code into YAML workflow definition.

**Comment by krancour** [comment]
> I think it’s not the right time

Note that no specific timeline has been established for 2.0. We're lumping issues like these together with a 2.0 label since that major release (whenever it happens) affords us a rare opportunity to make breaking changes and other major architectural changes.

Also note that we're not necessarily trying to compete / play in the same space as the likes of Argo. Indeed, if we decouple the Brigade control plane from k8s, an Argo-based implementation of the layer that drives/executes builds could even be a reality in the future-- but that requires us to first do the hard work of building the right abstractions and decoupling from k8s where we can.

**Comment by danmx** [comment]
That makes sense. Using existing components could speed up and enrich the tool, [Argo Events](https://github.com/argoproj/argo-events) covers gateways, [Argo Workflows](https://github.com/argoproj/argo) solves almost everything from event bus to artifacts and probably there are other tools that could cover other things.

**Comment by krancour** [comment]
This is well covered by the [2.0 proposal](https://github.com/brigadecore/brigade/tree/master/2.0-PROPOSAL), which has been ratified and is now guiding the 2.0 development effort. It probably doesn't make sense to track this as a discrete issue anymore.
